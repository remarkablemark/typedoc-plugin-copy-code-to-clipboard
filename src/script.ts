/**
 * Calls the plugin script.
 */
export default function init() {
  const text = 'Copy';
  const template = document.createElement('button');
  template.type = 'button';
  template.ariaLabel = 'Copy code to clipboard';
  template.innerText = text;
  document.querySelectorAll('pre').forEach((pre) => {
    const button = template.cloneNode(true) as HTMLButtonElement;
    button.addEventListener('click', function () {
      /* istanbul ignore next */
      const code = pre.querySelector('code')?.innerText;
      /* istanbul ignore if */
      if (code) {
        navigator.clipboard.writeText(code.trim());
      }
      button.innerText = 'Copied';
      setTimeout(() => {
        button.blur();
        button.innerText = text;
      }, 2e3);
    });
    pre.appendChild(button);
  });
}
