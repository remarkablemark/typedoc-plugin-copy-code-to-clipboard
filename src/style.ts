export default `
pre {
  position: relative;
}
pre > button {
  background: var(--code-background);
  border-radius: 0.8em;
  border: 0.1em solid var(--color-text-aside);
  cursor: pointer;
  height: 30px;
  opacity: 0;
  position: absolute;
  right: 5px;
  top: 5px;
  transition: all 0.2s ease-out;
  user-select: none;
}
pre:hover > button {
  opacity: 0.5;
}
pre > button:hover {
  opacity: 1;
}
pre > button:active, pre > button:focus {
  background: var(--color-background);
  opacity: 1;
  outline: 0;
}
`;
