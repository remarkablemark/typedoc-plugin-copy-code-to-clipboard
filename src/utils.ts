/**
 * Transforms function into Immediately Invoked Function Expression.
 *
 * @param func - Function.
 * @returns - IIFE.
 */
export function iife(func: () => void): string {
  const code = [func.toString(), `${func.name}();`].join('');
  return `(function(){${code}})();`;
}
