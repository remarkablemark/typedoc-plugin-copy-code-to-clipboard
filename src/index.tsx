import semver from 'semver';
import { Application } from 'typedoc';
import { JSX } from 'typedoc';

import script from './script';
import style from './style';
import { iife } from './utils';

/**
 * Loads the default TypeDoc plugin [hook](https://github.com/TypeStrong/typedoc/blob/master/internal-docs/custom-themes.md#hooks-v0228).
 *
 * @param app - Reference to the application that is loading the plugin.
 */
export function load(app: Application): void {
  // https://github.com/TypeStrong/typedoc/issues/2153
  const hasNativeSupport = semver.gte(Application.VERSION, '0.24.5');

  app.renderer.hooks.on('body.end', () => (
    <>
      <style>
        <JSX.Raw html={style.trim()} />
      </style>

      {!hasNativeSupport && (
        <script>
          <JSX.Raw html={iife(script)} />
        </script>
      )}
    </>
  ));
}
