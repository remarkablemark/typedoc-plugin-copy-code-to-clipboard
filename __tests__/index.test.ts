import { Application } from 'typedoc';

import { load } from '../src/index';

jest.mock('../src/script', () => 'script');
jest.mock('../src/style', () => 'style');

const { VERSION } = Application;

afterAll(() => {
  Application.VERSION = VERSION;
});

describe('load', () => {
  it.each(['0.24.4', '0.24.5'])(
    'renders markup on "body.end" for typedoc v%s',
    (version) => {
      Application.VERSION = version;
      const app = {
        renderer: {
          hooks: {
            on: jest.fn(),
          },
        },
      };
      expect(load(app as unknown as Application)).toBe(undefined);
      expect(app.renderer.hooks.on).toBeCalledTimes(1);
      expect(app.renderer.hooks.on).toBeCalledWith(
        'body.end',
        expect.any(Function)
      );
      expect(app.renderer.hooks.on.mock.calls[0][1]()).toMatchSnapshot();
    }
  );
});
