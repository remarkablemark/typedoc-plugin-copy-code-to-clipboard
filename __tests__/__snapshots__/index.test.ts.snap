// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load renders markup on "body.end" for typedoc v0.24.4 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "props": {
            "html": "style",
          },
          "tag": [Function],
        },
      ],
      "props": null,
      "tag": "style",
    },
    {
      "children": [
        {
          "children": [],
          "props": {
            "html": "(function(){scriptundefined();})();",
          },
          "tag": [Function],
        },
      ],
      "props": null,
      "tag": "script",
    },
  ],
  "props": null,
  "tag": Symbol(),
}
`;

exports[`load renders markup on "body.end" for typedoc v0.24.5 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "props": {
            "html": "style",
          },
          "tag": [Function],
        },
      ],
      "props": null,
      "tag": "style",
    },
    false,
  ],
  "props": null,
  "tag": Symbol(),
}
`;
